src/preprocessors/vue.ts

```diff
@@ -35,3 +37,60 @@ export function vuePreprocessor(code: string, options: PrettierOptions) {
         }
     }
 }
+
+function isTS(lang?: string) {
+    return lang === 'ts' || lang === 'tsx';
+}
+
+/**
+ * Configures correct babel plugins, sorts imports in a script or setupScript,
+ * and replaces that script/setupScript within the original code
+ *
+ * Much of this was adapted from https://github.com/vuejs/vue/blob/49b6bd4264c25ea41408f066a1835f38bf6fe9f1/packages/compiler-sfc/src/compileScript.ts#L118-L134
+ *
+ * @param param0 a script or setupScript
+ * @param code Source code of the file
+ * @param options Prettier options
+ * @returns Original code with sorted imports in the script provided
+ */
+function sortScript(
+    { content, lang }: { content: string; lang?: string },
+    code: string,
+    options: PrettierOptions,
+) {
+    const { importOrderParserPlugins = [] } = options;
+    let pluginClone = [...importOrderParserPlugins];
+    const newPlugins: ImportOrderParserPlugin[] = [];
+
+    if (!isTS(lang) || lang === 'tsx') {
+        newPlugins.push('jsx');
+    } else {
+        // Remove jsx if typescript and not tsx
+        pluginClone = pluginClone.filter((p) => p !== 'jsx');
+    }
+
+    newPlugins.push(...pluginClone);
+
+    if (isTS(lang)) {
+        if (!hasPlugin(newPlugins, 'typescript')) {
+            newPlugins.push('typescript');
+        }
+
+        if (
+            !hasPlugin(newPlugins, 'decorators') &&
+            !hasPlugin(newPlugins, 'decorators-legacy')
+        ) {
+            newPlugins.push('decorators-legacy');
+        }
+    }
```

Comment:
I’m uncomfortable with injecting `decorators`, because I’ve seen people fork every plugin. I worry that we’ll always be injecting decorators-legacy.

- https://babeljs.io/docs/babel-plugin-proposal-decorators

I’d be less uncomfortable if you were to shift this logic into the conditional-injection of the `‘typescript’` plugin. (Move the latter if-block to line 77).

I’d be least uncomfortable if this logic only happened if users omitted `importOrderParserPlugins` entirely. Then we definitely have permission to do automatic behavior. — If users are passing us plugins, I worry that manipulating it could only lead to more, painful errors.

Reply:
One problem we have is that we always inject `typescript` and `jsx`, and there's no way to know whether it was us or the user who did it.  I'd like to get to the point where we can remove those defaults and be smart about when we default to various plugins, but I'm not all the way there yet.  I'll see what it would take to do that as part of 4.0.

As for the decorators specifically, that was something I copied straight from vue (see https://github.com/vuejs/vue/commit/326d24a4e42b4f4fd243b36037e88b71963d963d).  I'm not a vue user, so I'm not sure why it's important to include a decorator plugin, but I think it might be related to the older class-based component definition in Vue 2: https://v2.vuejs.org/v2/guide/typescript.html#Class-Style-Vue-Components.

So, I'm happy to remove the injection of the decorators plugin.  Are you okay with the rest of it staying as-is?